### PR TITLE
Polish text in the "Scripting" section

### PR DESCRIPTION
**Pull request by d9n**
None of the changes in this CL change the meaning or spirit of the
underlying text. Instead, my goal was to tidy up miscellaneous issues
and make the text flow a little more smoothly.

### CODE REVIEW

**Comment by d9n** [comment]
Hey there!

Note that unlike my previous doc pull requests, I feel like I needed to modify this article a little more than usual to clarify things that I, at least, found I wanted a little more information about. I did my best to figure things out by Googling and experimentation. However, since I'm not a regular Godot user yet, **please double check if I wrote anything that's wrong or accidentally misleading**. If you push back on any section, I'll be happy to take a second look.

Thanks!

File: learning/step_by_step/scripting.rst [hunk 1]
```diff
@@ -7,75 +7,86 @@ Introduction
...
-fragmented into *complementary* scenes (not individual ones). Scenes
-complement each other, instead of being separate. There will be plenty
-of examples about this later on, but it's very important to remember it.
-
-For those with a good amount of programming expertise, this means a
-different design pattern to MVC. Godot promises efficiency at the
-expense of dropping the MVC habits, which are replaced by the *scenes as
-a complement* pattern.
-
-Godot also uses the `extend <http://c2.com/cgi/wiki?EmbedVsExtend>`__
```
Comment (d9n) [inline learning/step_by_step/scripting.rst]: Note that I moved this "extend" section into the GDScript section, where I felt it was more natural. It's also possible this paragraph could be entirely removed and I don't think most new readers would miss it, but in I'm not planning to delete text with any of my text polish suggestions.

File: learning/step_by_step/scripting.rst [hunk 2]
```diff
@@ -7,75 +7,86 @@ Introduction
...
+in this direction, adding tools that try to reduce the amount of code that needs
+to be written rather than eliminating it. The engine can provide many general
+solutions, while the developer can use code to accomplish specific tasks.
+
+Godot has embraced this goal from the beginning, and it has influenced many of
+its design decisions. First and foremost is the scene system. This system has
+many benefits, but fundamentally its goal is to relieve programmers from the
+responsibility of having to implement an overall architecture.
+
+As discussed in the previous section, when designing games using scenes, you
```
Comment (d9n) [inline learning/step_by_step/scripting.rst]: This is another paragraph that you could potentially remove. I'm not sure what the point is in this section of emphasizing that scenes are complementary, in a tutorial that's about scripting. But I'm happy to leave it in as well -- up to you.

File: learning/step_by_step/scripting.rst [hunk 3]
```diff
@@ -7,75 +7,86 @@ Introduction
...
+easier to learn, faster to write, no compilation, etc.) is balanced with a
+performance penalty. However, keep in mind that most critical code is already
+written in C++ in the engine (vector ops, physics, math, indexing, etc.), which
+results in more than sufficient performance for most types of games.
 
 In any case, if more performance is required, critical sections can be
-rewritten in C++ and exposed transparently to the script. This allows
-the replacement of a GDScript class with a C++ class without altering
-the rest of the game.
+rewritten in C++ and registered with Godot, which in turn exposes them to all
```
Comment (d9n) [inline learning/step_by_step/scripting.rst]: I assume this section was previously alluding to http://docs.godotengine.org/en/stable/development/cpp/custom_modules_in_cpp.html  However, it's now a bit more vague since Godot is introducing GDNative. Therefore, I tried to make it a bit clearer. Let me know if I missed the mark.

File: learning/step_by_step/scripting.rst [hunk 4]
```diff
@@ -91,129 +102,126 @@ look like the image below. You can set the text in the Inspector pane.
...
 menu:
 
 .. image:: /img/add_script.png
 
-The script creation dialog will pop up. This dialog allows to select
-the language, class name, etc. GDScript does not use class names in
-script files, so that field is not editable. The script should inherit
-from "Panel" (as it is meant to extend the node, which is of Panel type,
-this is automatically filled).
+The script creation dialog will pop up. This dialog allows you to set the
```
Comment (d9n) [inline learning/step_by_step/scripting.rst]: Tried to reorganize this section a bit to add a bit more detail about the "Name" and "Inherits" fields. Let me know what you think

File: learning/step_by_step/scripting.rst [hunk 5]
```diff
@@ -91,129 +102,126 @@ look like the image below. You can set the text in the Inspector pane.
...
 
 The role of the script
 ~~~~~~~~~~~~~~~~~~~~~~
 
-A script adds behavior to a node. It is used to control the
-node functions as well as other nodes (children, parent, siblings, etc).
-The local scope of the script is the node (just like in regular
-inheritance) and the virtual functions of the node are captured by the
-script.
+A script adds behavior to a node. It is used to control how the node functions
```
Comment (d9n) [inline learning/step_by_step/scripting.rst]: Slight wording change here, please review.

File: learning/step_by_step/scripting.rst [hunk 6]
```diff
@@ -91,129 +102,126 @@ look like the image below. You can set the text in the Inspector pane.
...
 
 .. image:: /img/signals.png
 
-In any case, at this point it is clear that we are interested in
-the "pressed" signal. Instead of using the visual interface, we will opt
-to code the connection.
+At this point, you could use the visual interface to hook up the "pressed"
+signal by double clicking on it and selecting a target node that already has a
+script attached to it. But for the sake of learning, we're going to code up the
+connection manually.
```
Comment (hubbyist) [inline learning/step_by_step/scripting.rst]: Order of this two sentences may be swapped, since first one is extra information for future tutorials.

File: learning/step_by_step/scripting.rst [hunk 7]
```diff
@@ -7,75 +7,86 @@ Introduction
...
-a complement* pattern.
-
-Godot also uses the `extend <http://c2.com/cgi/wiki?EmbedVsExtend>`__
-pattern for scripting, meaning that scripts extend from all the
-available engine classes.
+without programming. This appeals to many people who would love to be able to
+make a game without having to learn how to code. Furthermore, in many game
+companies, developers write custom tools that give their game designers more
+control over their game's flow. It's tempting, therefore, to imagine that
+someone could could write the ultimate tool that just works for any game type.
```
Comment (hubbyist) [inline learning/step_by_step/scripting.rst]: last sentence feels unnatural.

File: learning/step_by_step/scripting.rst [hunk 8]
```diff
@@ -7,75 +7,86 @@ Introduction
...
+
+Many products have been shipped promising such a no-programming environment,
+but the results often fall short of expectations. The final projects they
+produce regularly end up being too complex or require solutions that are too
+inefficient compared to what could have been accomplished with traditional code.
+
+In contrast, we believe that programming is a critical part of game development
```
Comment (hubbyist) [inline learning/step_by_step/scripting.rst]: "In contrast, we believe that" has a feeling of marketing phrase for a cheap product IMO.

File: learning/step_by_step/scripting.rst (hunk 2 above)
Comment (hubbyist) [inline learning/step_by_step/scripting.rst]: If "it's very important to remember it" it must stay I think.

File: learning/step_by_step/scripting.rst [hunk 9]
```diff
@@ -91,129 +102,126 @@ look like the image below. You can set the text in the Inspector pane.
...
-script files, so that field is not editable. The script should inherit
-from "Panel" (as it is meant to extend the node, which is of Panel type,
-this is automatically filled).
+The script creation dialog will pop up. This dialog allows you to set the
+language, class name, and other relevant options.
+
+Actually, in GDScript, the file itself represents the class, so in this case,
+the class name field is not editable.
+
+As the node we're creating the script on is of the Panel type, the "Inherits"
```
Comment (hubbyist) [inline learning/step_by_step/scripting.rst]: " creating the script on is of the Panel type" seems like a typo.

File: learning/step_by_step/scripting.rst (hunk 7 above)
Comment (d9n) [inline learning/step_by_step/scripting.rst]: Done. Reverted back to something more similar to the original sentence.

File: learning/step_by_step/scripting.rst (hunk 8 above)
Comment (d9n) [inline learning/step_by_step/scripting.rst]: Thanks. I've reverted to something closer to the original sentence.

The problem I was trying to solve is that the original sentence was

"As a result, programming is here to stay for a long time."

This is stated as fact, but we can't know 100% for sure. Something like "We feel" or "We think" seems required here.

Let me know what you think of this rewrite: " As a result, we think that programming is here to stay." Does that feel less marketing to you?

File: learning/step_by_step/scripting.rst (hunk 2 above)
Comment (d9n) [inline learning/step_by_step/scripting.rst]: Thanks! Leaving it in.

File: learning/step_by_step/scripting.rst (hunk 9 above)
Comment (d9n) [inline learning/step_by_step/scripting.rst]: Not quite a typo but yes, it can be hard to parse. Rewrote it to improve the flow.

File: learning/step_by_step/scripting.rst [hunk 10]
```diff
@@ -91,129 +101,126 @@ look like the image below. You can set the text in the Inspector pane.
...
+The editor provides an interface for connecting signals to your scripts. You
+can access this by selecting the node in the scene tree and then selecting the
+"Node" tab. Next, make sure that you have "Signals" selected.
 
 .. image:: /img/signals.png
 
-In any case, at this point it is clear that we are interested in
-the "pressed" signal. Instead of using the visual interface, we will opt
-to code the connection.
+At this point, you could use the visual interface to hook up the "pressed"
```
Comment (d9n) [inline learning/step_by_step/scripting.rst]: Sorry I appear to have lost old comments by doing a force push. Earlier here, you said:

"Order of this two sentences may be swapped, since first one is extra information for future tutorials."

However, the reason I used this order was the transition from the previous paragraph and image (which talk about "signals" in the visual UI" to the next paragraph (where we have the user start writing code). Therefore, I feel like changing the order of these sentences doesn't quite work. Did I miss something?

File: learning/step_by_step/scripting.rst [hunk 11]
```diff
@@ -91,129 +101,126 @@ look like the image below. You can set the text in the Inspector pane.
...
+To edit the script, select either of the highlighted buttons. This will bring
+you to the script editor where an existing template will be included by default:
 
 .. image:: /img/script_template.png
 
 There is not much in there. The "_ready()" function is called when the
-node (and all its children) entered the active scene. (Remember, it's
-not a constructor, the constructor is "_init()").
+node (and all its children) enter the active scene. (Note: "_ready()" is not
+the script's constructor; the constructor is "_init()").
```
Comment (Noshyaar) [inline learning/step_by_step/scripting.rst]: > *script's* constructor

It sounds strange to me (non-native speaker). It sounds like a script gets constructed when in fact it's an Object.

**Comment by akien-mga** [comment]
There's a conflict to solve after I've merged #425, sorry about that.

File: learning/step_by_step/scripting.rst (hunk 11 above)
Comment (d9n) [inline learning/step_by_step/scripting.rst]: No, that's fair. I reverted to "a constructor"

**Comment by d9n** [comment]
@akien-mga I resolved the merge conflict. Note that most of the changes made in #425 were also made by this CL, which was nice. I ended up using the text from that CL for the section on scenes being complementary. I also removed the hyphen from "dynamically-typed" since I left that as a comment on that CL, but if someone feels strongly about keeping the hyphen, I can put it back.

File: learning/step_by_step/scripting.rst [hunk 12]
```diff
@@ -7,75 +7,84 @@ Introduction
...
-goal. The first, and most important, is the scene system. The aim of it is
-not obvious at first, but is understandable later on. That is, it relieves
-programmers from the responsibility of designing and structuring code.
-
-When designing games using the scene system, the whole project is
-fragmented into *complementary* scenes (not individual ones). Scenes
-complement (i.e. help) each other, instead of being separate. There will be
-plenty of examples of this later on, but it's very important to remember it.
-
-For those with a good amount of programming expertise, this means a
```
Comment (d9n) [inline learning/step_by_step/scripting.rst]: Note to reviewers: I deleted this section. Please confirm.

As a programmer with a decent amount of programming expertise, I would say that I would not think about using MVC in game dev at all. (The entity-component pattern is at the heart of many games instead, I think) Maybe for UI, but even then, there are plenty of other patterns as well, e.g. MVVM. I guess my thought was, I don't really see how this paragraph helps experienced devs, who might just get stuck here and argue about the details. Meanwhile inexperienced or non-devs may be totally unaware of what MVC even is and this would only serve to distract them. 

Overall, I think this paragraph isn't necessary because I instead emphasized:

"... fundamentally [Godot's] goal is to relieve programmers from the responsibility of having to implement an overall architecture." 

which I believe captures the spirit of the original text. What do you think?

File: learning/step_by_step/scripting.rst (hunk 10 above)
Comment (hubbyist) [inline learning/step_by_step/scripting.rst]: I missed the previous sentences that are explaining the same situation in above paragraph, my comment was not accurate, so deleted it my self.

File: learning/step_by_step/scripting.rst (hunk 12 above)
Comment (hubbyist) [inline learning/step_by_step/scripting.rst]: "architecture" is fitting I think.

Comment (NathanLovato) [inline learning/step_by_step/scripting.rst]: You can find games and engines that rely on MVC, but it's not the most common approach indeed. It's good this way

**Comment by NathanLovato** [comment]
As a side-note, I'm tasked with rewriting the step-by-step section entirely, so please don't spend too much time on it right now. We had a meeting for the 3.0 docs 2 days ago. We're going to set a precise target audience and give it a more consistent format and style.  Your changes are good for the purpose of this article, but if we redo the entire section as a coherent series, there will likely be more changes. Especially in the writing style. There are lots of words we can remove to make the text easier to read.
But rest assured I'll do my best to retain your valuable improvements.

**Comment by d9n** [comment]
Thanks for being such a friendly, welcoming community!

Sounds good, Nathan. Actually my original intention was just to clean minor
grammar issues, mostly as a way to feel like I'm giving back while reading
through the docs anyway, but I got drawn into this section because there
was so much interesting going on. I've since noticed that the team is going
through the docs that and polishing grammar as well, so I'll refrain from
further edits. Honestly you can reject all my edits if there's no room for
them in the next version of the docs, I don't have any strong feelings
about my changes going in or not. Much happier if someone on the team who
is closer to the project is doing it, for sure!

**Comment by akien-mga** [comment]
Well if you're doing changes as you go through the docs yourself, feel free to continue. @NathanLovato will be working on the docs for the upcoming 3.0 version, but the changes you made so far should and will be backported to the stable 2.1 branch, so they will improve the 2.1 documentation even if they ended up being completely rewritten for 3.0.

File: learning/step_by_step/scripting.rst (hunk 5 above)
Comment (akien-mga) [inline learning/step_by_step/scripting.rst]: Seems good.
